### PR TITLE
[OpenMP] Predetermined rule for sequential loop index

### PR DESCRIPTION
**Pull request by ichoyjx**
This commit implements rule:
A loop iteration variable for a sequential loop in a parallel or
task generating construct is private in the innermost such construct
that encloses the loop.

A Simple example:
```
  i = -1                    <== Scope 0
  j = -1
  !$omp parallel            <== Scope 1
  print *,i,j      <-- both are shared (Scope 0)
  !$omp parallel            <== Scope 2
  print *,i,j      <-- a) i is shared (Scope 0), j is private (Scope 2)
  !$omp do                  <== Scope 3
  do i=1, 10       <-- i is private (Scope 3)
     do j=1, 10    <-- b) j is private (Scope 2, not 3!)
     enddo
  enddo
  print *,i,j      <-- c) i is shared (Scope 0), j is private (Scope 2)
  !$omp end parallel
  print *,i,j      <-- both are shared (Scope 0)
  !$omp end parallel
  print *,i,j      <-- both are shared (Scope 0)
end
```

The basic idea is when visiting the `DoConstruct` node within an OpenMP
construct, if the do-loop is not associated (like `i` loop is associated
with `!$omp do`) AND the do-loop is in the parallel/task generating
construct, resolve the loop index to be private to that innermost construct.

In the above example, `j` loop is not associated (then it is sequential) and
the innermost parallel/task generating construct that encloses the `j` loop
is the `parallel` construct marked with `<== Scope 2`, so `j` is private
to that construct. To do that, I also need to change the prototype of those
`ResolveOmp*` functions to allow specifiying the `scope` because the new
symbol for `j` should be created in Scope 2 and all the `symbol` field of
`Name j` in that `parallel` construct should be fixed, such as c).

### CODE REVIEW

**Comment by ichoyjx** [comment]
> Everything builds, tests, and looks good to me. I hesitate to approve since I don't really understand the changes.

Thanks Pete!!

File: lib/semantics/resolve-names.cpp [hunk 1]
```diff
@@ -6217,19 +6274,20 @@ Symbol *OmpAttributeVisitor::ResolveOmpCommonBlockName(
 }
 
 void OmpAttributeVisitor::ResolveOmpObjectList(
-    const parser::OmpObjectList &ompObjectList, Symbol::Flag ompFlag) {
+    const parser::OmpObjectList &ompObjectList, Symbol::Flag ompFlag,
+    Scope &scope) {
```
Comment (tskeith) [inline lib/semantics/resolve-names.cpp]: Why the change to pass in scope when it is always currScope() anyway?

File: lib/semantics/check-omp-structure.h [hunk 2]
```diff
@@ -88,6 +88,8 @@ static constexpr OmpDirectiveSet simdSet{
     OmpDirective::TASKLOOP_SIMD,
     OmpDirective::TEAMS_DISTRIBUTE_PARALLEL_DO_SIMD,
     OmpDirective::TEAMS_DISTRIBUTE_SIMD};
+static constexpr OmpDirectiveSet taskGeneratingSet{
+    OmpDirectiveSet{OmpDirective::TASK} | taskloopSet};
```
Comment (tskeith) [inline lib/semantics/check-omp-structure.h]: It looks like taskGeneratingSet and similar sets can be defined in the .cpp file instead of the .h file.

File: lib/semantics/resolve-names.cpp [hunk 3]
```diff
@@ -6051,6 +6063,53 @@ bool OmpAttributeVisitor::Pre(const parser::OpenMPLoopConstruct &x) {
...
+  for (auto it{ompContext_.rbegin()}; it != ompContext_.rend(); ++it) {
+    if (parallelSet.test(it->directive) ||
+        taskGeneratingSet.test(it->directive)) {
+      symbol = ResolveOmp(iv, Symbol::Flag::OmpPrivate, it->scope);
+      targetIt = it;
+      break;
+    }
+  }
+  targetIt++;
+  if (symbol) {
```
Comment (tskeith) [inline lib/semantics/resolve-names.cpp]: The increment should be inside the `if` because that is where `targetIt` is used. And `targetIt` doesn't need to be initialized.

File: lib/semantics/resolve-names.cpp [hunk 4]
```diff
@@ -6051,6 +6062,53 @@ bool OmpAttributeVisitor::Pre(const parser::OpenMPLoopConstruct &x) {
...
+      AddToContextObjectWithDSA(*symbol, Symbol::Flag::OmpPrivate, *it);
+    }
+  }
+}
+
+// 2.15.1.1 Data-sharing Attribute Rules - Predetermined
+//   - A loop iteration variable for a sequential loop in a parallel
+//     or task generating construct is private in the innermost such
+//     construct that encloses the loop
+bool OmpAttributeVisitor::Pre(const parser::DoConstruct &x) {
```
Comment (jeanPerier) [inline lib/semantics/resolve-names.cpp]: How the "sequential" aspect is determined here ? Should it be tested that the `parser::DoConstruct` belongs to a `parser::OpenMPLoopConstruct` ?

File: lib/semantics/resolve-names.cpp [hunk 5]
```diff
@@ -6051,6 +6062,53 @@ bool OmpAttributeVisitor::Pre(const parser::OpenMPLoopConstruct &x) {
...
+  }
+  if (symbol) {
+    targetIt++;
+    symbol->set(Symbol::Flag::OmpPreDetermined);
+    iv.symbol = symbol;  // adjust the symbol within region
+    for (auto it{ompContext_.rbegin()}; it != targetIt; ++it) {
+      AddToContextObjectWithDSA(*symbol, Symbol::Flag::OmpPrivate, *it);
+    }
+  }
+}
```
Comment (jeanPerier) [inline lib/semantics/resolve-names.cpp]: Don't you also need to fix all `parse::Name` symbols that were visited before it was determined their symbols is actually private ?
You may need a second pass here on all the statements inside the inner parallel or task construct. If so, I am not sure though what would be the best/easier between doing so every-time you need to fix a symbol, or make two passes overall, one to determine what symbol needs to be fixed in which scope, and another one later to fix all the relate `parser::Name` symbols.

Comment (ichoyjx) [reply lib/semantics/resolve-names.cpp]: Yes, it's a problem & I've been thinking about the solutions since last week. Similar to the comment you made for `omp-symbol08.f90`, this is the issue listed as TODO item a) in this PR description. Two potential solutions:
   * two passes (then I need to make sure that each OpenMP msg is emitted only for once)
   * maintain a list or map (or some other DSes) for all the `Names` processed and go back to fix those

I think that I lean towards the second one.

File: lib/semantics/resolve-names.cpp (hunk 4 above)
Comment (ichoyjx) [reply lib/semantics/resolve-names.cpp]: No. The `sequential` here means that do-loop is within OpenMP construct but not associated with `OpenMPLoopConstruct`. Like:
```
!$omp parallel do
do i = 1, 10   <-- associated
   do j = 1, 10   <-- sequential
    ...
   end do
enddo
```

Comment (jeanPerier) [reply lib/semantics/resolve-names.cpp]: How do you make the difference between loops that are associated to `OpenMPLoopConstruct` or not here ?

File: lib/semantics/resolve-names.cpp (hunk 5 above)
Comment (jeanPerier) [reply lib/semantics/resolve-names.cpp]: Sorry, missed the TODO in PR description. The map/list/vector/... of `Name`s may be an interesting option indeed.

File: lib/semantics/resolve-names.cpp (hunk 4 above)
Comment (ichoyjx) [reply lib/semantics/resolve-names.cpp]: Through line 6102 below, if the loop index is already determined as `OmpPredetermined`, then ignore.

**Comment by ichoyjx** [comment]
Note that the latest commit added the second `Walk` of `OmpAttributeVisitor` if there is no error during the first `Walk`, this resolves the issue a).

File: lib/semantics/resolve-names.cpp [hunk 6]
```diff
@@ -6076,6 +6095,53 @@ bool OmpAttributeVisitor::Pre(const parser::OpenMPLoopConstruct &x) {
...
+  std::vector<OmpContext>::reverse_iterator targetIt;
+  for (auto it{ompContext_.rbegin()}; it != ompContext_.rend(); ++it) {
+    if (parallelSet.test(it->directive) ||
+        taskGeneratingSet.test(it->directive)) {
+      symbol = ResolveOmp(iv, Symbol::Flag::OmpPrivate, it->scope);
+      targetIt = it;
+      break;
+    }
+  }
+  if (symbol) {
```
Comment (tskeith) [inline lib/semantics/resolve-names.cpp]: I think this is a better way to write this:
```
  auto targetIt{ompContext_.rbegin()};
  for (;; ++targetIt) {
    if (targetIt == ompContext_.rend()) {
      return;
    }
    if (parallelSet.test(targetIt->directive) ||
        taskGeneratingSet.test(targetIt->directive)) {
      break;
    }
  }
  if (Symbol *symbol{ResolveOmp(iv, Symbol::Flag::OmpPrivate, targetIt->scope)}) {
```

File: lib/semantics/resolve-names.cpp [hunk 7]
```diff
@@ -5771,6 +5782,14 @@ bool ResolveNamesVisitor::Pre(const parser::ProgramUnit &x) {
...
   ResolveExecutionParts(root);
   OmpAttributeVisitor{context(), *this}.Walk(x);
+  if (!context().AnyFatalError()) {
+    // The data-sharing attribute of the loop iteration variable for a
+    // sequential loop (2.15.1.1) can only be determined when visiting
+    // the corresponding DoConstruct, a second walk is to adjust the
+    // symbols for all the data-refs of that loop iteration variable
+    // prior to the DoConstruct.
+    OmpAttributeVisitor{context(), *this}.Walk(x);
+  }
```
Comment (tskeith) [inline lib/semantics/resolve-names.cpp]: Can you extract this into a separate function so that this function shows the high-level order of processing a program unit?

File: lib/semantics/resolve-names.cpp (hunk 6 above)
Comment (ichoyjx) [reply lib/semantics/resolve-names.cpp]: thank you for the detailed suggestion! I've made the changes.

**Comment by jeanPerier** [comment]
LGTM
